# Build a Deep Learning model assignement


<a id="item31"></a>


## Data preparation


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [1]:
import pandas as pd
import numpy as np

### We start by reading and understanding data...


In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
print("Data Header")
display(concrete_data.head())
print("\n","*"*70,"\n")
print("Data Summary")
display(concrete_data.describe())
print("\n","*"*70,"\n")
print("Missing values")
concrete_data.isnull().sum()

Data Header


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30


 ********************************************************************** 

Data Summary


Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000


 ********************************************************************** 

Missing values


Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


The data looks very clean and is ready to be used to build our model.


### We split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [4]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
predictors.head()


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

<a id="item2"></a>


In [5]:
target = concrete_data['Strength'] # Strength column
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [6]:
n_cols = predictors.shape[1] # number of predictors

## PART D

### Assignement

Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

#### Lets start with normalizing data

In [7]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [8]:
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [9]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

# build the model
model = regression_model()

#### 1. Randomly split the normalized data into a training and test sets by holding 30% with sklearn

<a id='item34'></a>


In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=44)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/sklearn/utils/validation.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  LARGE_SPARSE_SUPPORTED = LooseVersion(scipy_version) >= '0.14.0'


Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


#### 2. Train the model on the training data using 50 epochs.

In [11]:
model.fit(X_train, y_train, epochs=50, verbose=1)



Epoch 1/50


2022-07-21 11:56:19.052934: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2022-07-21 11:56:19.059962: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2494065000 Hz
2022-07-21 11:56:19.060584: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x564ff2160f40 executing computations on platform Host. Devices:
2022-07-21 11:56:19.060625: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2022-07-21 11:56:19.198741: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line fl

721/721 [==============================] - 0s 535us/step - loss: 1600.6304
Epoch 2/50
721/721 [==============================] - 0s 134us/step - loss: 1585.8774
Epoch 3/50
721/721 [==============================] - 0s 90us/step - loss: 1575.8655
Epoch 4/50
721/721 [==============================] - 0s 94us/step - loss: 1565.9732
Epoch 5/50
721/721 [==============================] - 0s 106us/step - loss: 1553.1794
Epoch 6/50
721/721 [==============================] - 0s 89us/step - loss: 1534.3667
Epoch 7/50
721/721 [==============================] - 0s 81us/step - loss: 1506.9737
Epoch 8/50
721/721 [==============================] - 0s 68us/step - loss: 1466.1737
Epoch 9/50
721/721 [==============================] - 0s 80us/step - loss: 1408.7824
Epoch 10/50
721/721 [==============================] - 0s 83us/step - loss: 1329.4506
Epoch 11/50
721/721 [==============================] - 0s 81us/step - loss: 1228.3015
Epoch 12/50
721/721 [==============================] - 0s 83us/step - l

#### 3. Evaluate the model on the test data and compute the mean squared error

##### 3.1. Evaluation with keras

In [12]:
print("loss = ", model.evaluate(X_test, y_test))

309/309 [==============================] - 0s 127us/step
loss =  156.47433116135088


##### 3.2. Evaluation with sklearn

In [13]:
# predictions
y_pred = model.predict(X_test)

from sklearn.metrics import mean_squared_error

print("loss = ", mean_squared_error(y_test, y_pred))

loss =  156.47433336154512


#### 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.


In [ ]:
mse_list=[]
for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=50, verbose=1)
    loss = model.evaluate(X_test, y_test)
    mse_list.append(loss)
   

Epoch 1/50
721/721 [==============================] - 0s 86us/step - loss: 146.0022
Epoch 2/50
721/721 [==============================] - 0s 84us/step - loss: 145.4387
Epoch 3/50
721/721 [==============================] - 0s 80us/step - loss: 143.7388
Epoch 4/50
721/721 [==============================] - 0s 79us/step - loss: 142.6781
Epoch 5/50
721/721 [==============================] - 0s 83us/step - loss: 142.0217
Epoch 6/50
721/721 [==============================] - 0s 80us/step - loss: 141.0157
Epoch 7/50
721/721 [==============================] - 0s 80us/step - loss: 140.1479
Epoch 8/50
721/721 [==============================] - 0s 82us/step - loss: 139.6694
Epoch 9/50
721/721 [==============================] - 0s 81us/step - loss: 138.6706
Epoch 10/50
721/721 [==============================] - 0s 81us/step - loss: 137.9903
Epoch 11/50
721/721 [==============================] - 0s 88us/step - loss: 137.3068
Epoch 12/50
721/721 [==============================] - 0s 100us/step - los

In [ ]:
print("We reapeat the 3 first steps and create a list including", len(mse_list), "MSE")

We reapeat the 3 first steps and create a list including 50 MSE


#### 5.Report the mean and the standard deviation of the mean squared errors.

##### 5.1. Mean of the 50 mean squared errors

In [ ]:
print("Mean of the mean squared errors = ", np.mean(mse_list))

Mean of the mean squared errors =  29.17747606592271


##### 5.2. Standard deviation of the 50 mean squared errors

In [ ]:
print("The standard deviation of the mean squared errors = ", np.std(mse_list))

The standard deviation of the mean squared errors =  14.468549646527872


### How does the mean of the mean squared errors compare to that from Step B?

- Mean MSE in part-B = 42.88
- Mean MSE in part-B = 29.18

**The more complexe the network the better the model scores**




### Thank you !